contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,456 @@
+by different apiservers
+- clients has to be able to alter their watch subscribtions incrementally (it
+may not be implemented in the initial version though, but has to be designed)
+
+
+# Detailed design
+
+ As stated in above requirements, we need to make bulk operations work across
+different resource types (e.g. watch pod P and secret S within a single watch
+call). To clarify, the initial implementation doesn't need to support it, but
+it has to be possible to extend the implementation without changing the API.
+Spanning multiple resources, resources types or conditions will be more and more
+important for large number of watches. As an example, federation will be adding
+watches for every type it federates. With that in mind, bypassing aggregation
+at the resource type level and going to aggregation over objects with different
+resource types will allow us to more aggresively optimize in the future.
+
+ Moreover, out current REST API doesn't even offer an easy way to handle
+"multiple watches of a given type" within a single request. As a result, instead
+of inventing new top level pattern per type, we can introduce a new resource
+type that follows normal RESTful rules and solves even more generic problem
+of spanning multiple different resource types.
+
+ We will start with introducing a new dedicated API group:
+ ```
+ /apis/bulk.k8s.io/
+ ```
+ that underneath will have a completely separate implementation.
+
+ In this design, we will focus only on bulk list (get) and watch operations.
+Later, we would like to introduce new resources to support bulk create, update
+and delete operations, but that's not part of this design.
+
+ We will start with introducing `list` resource and supporting the following
+two operation:
+```
+POST /apis/bulk.k8s.io/v1/list <body defines filtering>
+POST /apis/bulk.k8s.io/v1/list?watch=1 <body defines filtering>
+```
+ We can't simply make this an http GET request, due to limitations of GET for
+the size (length) of the url (in which we would have to pass filter options).
+This makes the API look similar to our regular API.
+ Note that once we have list fully implemented, adding this kind of watch will
+be faily simple.
+
+ The main drawback of this approach is that it won't allow for dynamic altering
+of watch subscriptions (which we would like to also support).
+As a result, we need another API for watch that will also support incremental
+subscribtions. That's why we will also introduce the following operation:
+```
+websocket /apis/bulk.k8s.io/v1/list?watch=1
+```
+ The more detailed description of this protocol is in the followin subsection.
+
+ We consistency, we may also consider introducing websocket API for handling
+LIST requests, i.e.:
+```
+websocket /apis/bulk.k8s.io/v1/list
+```
+ where first client sends filter definition over the channel and then server
+sends back the response (all further writes from client are ignored). This
+would make the list API consistent with watch.
```

Comment:
Originally (pre HTTP2) I would have said, yes, we want to do this.  However, with HTTP2, it's not as useful.  I haven't thought a lot about this, but I'd say that a future evolution of the API would be more towards gRPC than towards a raw HTTP2 endpoint.  In gRPC, we'd have connection reuse, and you'd simply invoke an RPC endpoint to begin a watch on a remote service:

```
for _, selector := range selectors {
  stream, err := pods.Watch(WatchRequest{rv, selector})
}
...
```

If we had a gRPC endpoint, we wouldn't need bulk.  So whatever we do for bulk watch, we should keep constrained to solving the problem in a way that feels the most like bulk watch invocation.  Which is that we can start watches at any time, and then remove them.

List and Watch are often paired.  We model those for singular watch as `GET /resource?selector=...` and `GET /resource?watch=1&selectors=...`.

For bulk watch, we want them to retain their logical pairing.  However, bulk watch is very much like publish / subscribe.  So there's the open question whether we should do a separate list endpoint, or instead extend the bulk watch protocol to support LIST.  That would maintain the pairing (keeping LIST and WATCH similar) and also allow us to better support naive clients that want to do list/watch.

If we retroactively looked at what we do for single resource list watch, in the context of bulk watch, we'd allow you to upgrade `GET /resources?watch=1&selector=...` into a websocket.